### ISSUE TITLE
Explore AMD to ESM migration

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#160416]
The VS Code project uses AMD (asynchronous module definition) which is somewhat of a pre-runner of native JS modules (ESM). AMD is mostly dead technology and we should migrate away from it. Our TypeScript sources are written with ESM imports/exports but we took a few dependencies on AMD as a "runtime" and we need to understand what it takes to move towards ESM. 

High-level topics:

* we have a special AMD loader plugin that supports nls
* we also have a special plugin for CSS
* we have special bundle logic for both NLS and CSS dependencies
* we use `require.toUrl` to locate sibling resources, like images, this needs to migrate onto `import.meta.url`

Backlog Plan: https://github.com/microsoft/vscode/issues/160416#issuecomment-2269225658

Test Builds that run on ESM:

```
VS Code Build: 7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d

darwin: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/VSCode-darwin.zip
darwin-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/VSCode-darwin-arm64.zip
darwin-universal: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/VSCode-darwin-universal.zip
linux-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration-arm64-1723616523.tar.gz
linux-deb-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration_1.93.0-1723616603_arm64.deb
linux-deb-x64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration_1.93.0-1723617891_amd64.deb
linux-rpm-arm64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration-1.93.0-1723616690.el8.aarch64.rpm
linux-rpm-x64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration-1.93.0-1723617981.el8.x86_64.rpm
linux-snap-x64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration-x64-1723619959.snap
linux-x64: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/code-exploration-x64-1723616784.tar.gz
win32-arm64-user: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/VSCodeUserSetup-arm64-1.93.0-exploration.exe
win32-x64-user: https://vscode.download.prss.microsoft.com/dbazure/download/exploration/7c7401b2553c7188f1a634b5264fc3a6d3cb1e7d/VSCodeUserSetup-x64-1.93.0-exploration.exe
```

**Comment by lorand-horvath** [comment]
Huh, VSCode still using AMD...

**Comment by jasonwilliams** [comment]
Hey @jrieken where would this leave extensions? Would there be a path for them to export esm? I believe right now they still export to commonjs because electron does not support esm

**Comment by jrieken** [comment]
> Hey @jrieken where would this leave extensions? Would there be a path for them to export esm? I believe right now they still export to commonjs because electron does not support esm

@jasonwilliams For now extensions aren't affected by this and for the time being commonjs is how they roll. Tho, I am learning a ton and there will be a path for ESM extensions. We ignore this to keep this already large effort smaller. Also, if you wonder why this is complex: when extensions call `require('vscode')` something very special happens, there is no file for `vscode` but each extension gets an instance of a the API. This is only possible with commonjs loader tricks and we need to find similar ways in the ESM world.

**Comment by jrieken** [comment]
Yeah, import maps is our only viable option for this. Fingers crossed for Safari landing this in time for us

**Comment by zm-cttae** [comment]
If Safari doesn't support this, aren't we blocked by the few months it takes for browser updates to percolate to userland?

See Google Chrome 110 popularity vs Google Chrome 109 here: https://www.stetic.com/market-share/browser/

**Comment by jasonwilliams** [comment]
Looks like Safari have included import maps support in their latest release https://developer.apple.com/documentation/safari-release-notes/safari-16_4-release-notes

**Comment by jasonwilliams** [comment]
@jrieken I am curious how you are going to achieve this when Electron doesn't support ESM. 
https://github.com/electron/electron/issues/21457

Is this something you've considered?

**Comment by zm-cttae** [comment]
Doesn't [Node 16.14.2](https://github.com/ewanharris/vscode-versions#vs-code-versions/tree/a6485ab70915e56132663ca21a5672c120ab35b8) support it?

**Comment by Jack-Works** [comment]
https://github.com/guybedford/es-module-shims supports import map polyfill

**Comment by sualehasif** [comment]
Curious to the results of the exploration? What are the blocking issues for the migration?

**Comment by JustinGrote** [comment]
Electron now has support for this so the question will be when the newest Electron gets merged into vscode
![image](https://github.com/microsoft/vscode/assets/15258962/834dea83-b902-4a45-83a0-444d073331ee)

**Comment by SimonSiefke** [comment]
@jrieken if it's fine with you, I could work on moving the `build` folder to esm. 

Perhaps after that, some of the workers (`textMateTokenizationWorker`, `outputLinkComputer`, `editorWorkerService`) could be moved to esm in three steps: 
1. For each of the workers above, Instead of creating a Classic-Worker, create a Module-Worker 
2. Change `simpleWorker` so that it can not only `require`, but also `import` another file
3. Change the build process so that the final `textMateTokenizationWorker`, `outputLinkComputer` and `editorWorkerService` files are esm

**Comment by jrieken** [comment]
> @jrieken if it's fine with you, I could work on moving the build folder to esm.

Thanks for the offer but migration the build folder isn't on this plan. It uses commonjs, not AMD, and hence different cup of tea. Not saying it shouldn't be done but it's nothing pressing and would need team support first.

**Comment by jasonwilliams** [comment]
Hey @jrieken would love to know if you have any update at on this. It sounds like all the building blocks you need to make this happen have arrived. (Safari supporting import maps, Electron supporting ESM), it feels like there shouldn’t the any blockers to making this a reality now.

**Comment by JustinGrote** [comment]
@jasonwilliams The Electron version vscode uses still has to reach the version that supports ESM, that's a ways away still.

**Comment by JustinGrote** [comment]
As an update, the electron version is now on an ESM-supportable version, but the loader still needs to be fixed to enable it. This is no longer an external dependency now however.

https://github.com/microsoft/vscode-loader/issues/56

**Comment by metawrap-dev** [comment]
I will be very glad when this is resolved. I have a component of a project on hold until this can be resolved.

**Comment by jrieken** [comment]
Things for July

* - [ ] bring the `alex/esm` branch back into a good state
  *   - [x] run oss: `./scripts/code`
  *   - [ ] run oss: `./scripts/code-cli`
  *   - [ ] run oss: `./scripts/code-web`
  *   - [ ] run oss: `./scripts/code-server`
  *   - [x] tests: `./scripts/test`
  *   - [ ] tests: `yarn run test-browser`
  *   - [ ] tests: `yarn run test-node`
  *   - [ ] tests: `scripts/test-integration`
  *   - [ ] tests: `scripts/test-web-integration`
  *   - [ ] tests: `scripts/test-remove-integration`
* - [ ] Adopt electron support for ESM
* - [ ] Adopt `node:module#register` over server-loader.mjs (during dev we redirect to node_modules from the remote folder)

**Comment by bpasero** [comment]
`yarn test-node` works, but shows some interesting few errors when loading `jschardet`:

```
TypeError: Cannot read properties of undefined (reading 'iterator')
      at $jscomp.initSymbolIterator (evalmachine.<anonymous>:1:384)
      at $jscomp.makeIterator (evalmachine.<anonymous>:2:42)
      at evalmachine.<anonymous>:164:334
      at new c (evalmachine.<anonymous>:164:465)
      at evalmachine.<anonymous>:662:132
      at 42../constants (evalmachine.<anonymous>:662:395)
      at f (evalmachine.<anonymous>:5:32)
      at evalmachine.<anonymous>:5:66
      at 19../logger (evalmachine.<anonymous>:390:112)
      at f (evalmachine.<anonymous>:5:32)
      at evalmachine.<anonymous>:5:66
      at 1../src (evalmachine.<anonymous>:5:249)
      at f (evalmachine.<anonymous>:5:32)
      at a (evalmachine.<anonymous>:5:190)
      at DefineCall.callback (evalmachine.<anonymous>:5:218)
      at AMDModuleImporter.load (file:///Users/bpasero/Development/Microsoft/vscode-esm/out/vs/amdX.js:79:31)
```

**Comment by SimonSiefke** [comment]
@bpasero I encountered the same error. It seems the issue is that `this` is undefined at the top level of Ecmascript modules: 


In `jschardet.min.js`: 

![this](https://github.com/microsoft/vscode/assets/23744935/f4a58d04-2133-40be-8e76-0798aa14e0c4)


```js
var $jscomp = { scope: {}, global: this }, // this is undefined
	Symbol;
$jscomp.initSymbol = function () {
	$jscomp.global.Symbol || (Symbol = $jscomp.Symbol); // error occurs here
	$jscomp.initSymbol = function () {};
};
```



When accessing `$jscomp.global.Symbol` it is accessing `undefined.Symbol`. 



There is a PR here https://github.com/microsoft/vscode-loader/pull/58 that might fix it.

**Comment by bpasero** [comment]
Thanks for the insights, maybe I am missing something but in the `alex/esm` (https://github.com/microsoft/vscode/pull/166033) branch, we do not use the `vscode-loader` anymore I had thought 🤔

**Comment by SimonSiefke** [comment]
Thanks for the reply. I thought it might still be required, e.g. when an extension uses `require` in a webworker?

But I think you're also right that the same issue applies to the amdX loader.

Before, `this` was always a defined value. But that EcmaScript modules are used, `this` is undefined, causing an error when importing jschardet. 

A possible workaround could be to manually overwrite the jschardet code: 

```js
var $jscomp = { scope: {}, global: globalThis }, // change here
	Symbol;
$jscomp.initSymbol = function () {
	$jscomp.global.Symbol || (Symbol = $jscomp.Symbol);
	$jscomp.initSymbol = function () {};
};
```

**Comment by jrieken** [comment]
> Thanks for the reply. I thought it might still be required, e.g. when an extension uses require in a webworker?

We fake `require` in that case. It's a single fetch and doesn't support dependencies ;-)

**Comment by jrieken** [comment]
> It seems the issue is that this is undefined at the top level of Ecmascript modules:

@SimonSiefke The `amdX` module is actually a very simple AMD loader that we use to load our existing dependencies. This is a stepping stone so that the migration doesn't spread into all the places... When debugging the load-call, I see that `this` is defined and the error happens later when trying to access `Symbol` (which should also be defined...)

![Image](https://github.com/user-attachments/assets/82b07b2a-61bb-4047-b243-f2906e171de8)

I didn't get to the bottom of this yet, it is a really weird error. Things work when using the non-minified version which I'll do for now

**Comment by SimonSiefke** [comment]
@jrieken  you're right. It looks like `Symbol` is defined but it's running in a `vm` Context where accessing `Symbol.iterator` is not allowed because it is a different v8 context (or something like that?). 

It seems only the `jschardet.min.js` version has this extra `Symbol` code for supporting older browsers that don't support `for .. of` loops.

**Comment by Tyriar** [comment]
With https://github.com/xtermjs/xterm.js/pull/5092 xterm.js now ships an ESM module.

**Comment by bpasero** [comment]
Re `jschardet`: I opened https://github.com/aadsm/jschardet/pull/96 to try to get the project to update to a newer version of `google-closure-compiler` that they use for minification. I am not able to reproduce the issue anymore with newer versions. It is possible that maybe some polyfills (such as `for...of`) are not applied anymore when using the newer compiler versions.

For now I think its OK to use the non-minified release. Eventually we should consider to minify all our node module dependencies in our pipeline and not depend on how others do it.

**Comment by bpasero** [comment]
🚀 `main` is now capable of running as ESM, no more need for a branch!

**Developing locally from `main`:**
* `node migrate.mjs` (this will copy `src` to `src2` with ESM transformations)
* `yarn watch-esm` (this will compile from `src2` to `out`)
* `scripts/code.sh`

**Unit Tests:**
* `scripts/test-esm.sh`
* `yarn test-browser-esm`
* `yarn test-node-esm`

**Comment by bpasero** [comment]
And we are happy for people willing to test our VS Code exploration builds that run with ESM. These builds can be side-by-side installed to VS Code and will update daily like insiders:

* Windows (x64): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/VSCodeUserSetup-x64-1.93.0-exploration.exe)
* Windows (ARM): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/VSCodeUserSetup-arm64-1.93.0-exploration.exe)
* macOS (Universal): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/VSCode-darwin-universal.zip)
* Linux (x64, deb): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/code-exploration_1.93.0-1723716636_amd64.deb)
* Linux (ARM, deb): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/code-exploration_1.93.0-1723715483_arm64.deb)
* Linux (x64, RPM): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/code-exploration-1.93.0-1723716726.el8.x86_64.rpm)
* Linux (ARM, RPM): [Download](https://vscode.download.prss.microsoft.com/dbazure/download/exploration/06dde2f7a942b5daf61ed2ce4dcb13e6016402e8/code-exploration-1.93.0-1723715572.el8.aarch64.rpm)

Feel free to report any issues you see, thanks!

**Comment by bpasero** [comment]
We have finished the majority of this work and this issue is closed with 2 follow up issues:
* https://github.com/microsoft/vscode/issues/226399
* https://github.com/microsoft/vscode/issues/226260

Thanks for being patient with us 🙏
